File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1 for _ in range(len(nums))]
        suffix = [1 for _ in range(len(nums))]

        prefix[0] = nums[0]
        suffix[-1] = nums[-1]

        for i in range(1, len(nums)):
            prefix[i] = prefix[i - 1] * nums[i]

        for i in range(len(nums) - 1):
            suffix[len(nums) - i - 2] = suffix[len(nums) - i - 1] * nums[len(nums) - i - 2]

        result = []

        for i in range(len(nums)):
            if i == 0:
                result += [suffix[1]]
            if i == len(nums) - 1:
                result += [prefix[len(nums) - 2]]
            if i > 0 and i < len(nums) - 1:
                result += [suffix[i + 1] * prefix[i - 1]]

        return result

File: Data Structures & Algorithms/house-robber-ii/submission-0.java
class Solution {
    public static int rob(int[] nums) {
        Map<String, Integer> map = new HashMap<>();
        return accomplice(nums, 0, map, false);
    }

    private static int accomplice(int[] nums, int i, Map<String, Integer> map, boolean robbedFirstHouse) {
        if (i >= nums.length)
            return 0;

        if (map.containsKey(formatKey(i, robbedFirstHouse)))
            return map.get(formatKey(i, robbedFirstHouse));

        if (i == 0) {
            int result = Math.max(nums[i] + accomplice(nums, i + 2, map, true), accomplice(nums, i + 1, map, false));
            map.put(formatKey(i, robbedFirstHouse), result);
            return result;
        }

        if (i == nums.length - 1) {
            int result = Math.max(robbedFirstHouse ? 0 : nums[i] + accomplice(nums, i + 2, map, robbedFirstHouse),
                    accomplice(nums, i + 1, map, robbedFirstHouse));
            map.put(formatKey(i, robbedFirstHouse), result);
            return result;
        }

        int result = Math.max(nums[i] + accomplice(nums, i + 2, map, robbedFirstHouse),
                accomplice(nums, i + 1, map, robbedFirstHouse));
        map.put(formatKey(i, robbedFirstHouse), result);
        return result;
    }

    private static String formatKey(int i, boolean bool) {
        return i + "," + bool;
    }
}
